### ISSUE TITLE
How should Previous/Next buttons be supported?

### ISSUE THREAD

**Issue by terrill** [ableplayer/ableplayer#200]
There have been many requests over the years for "Previous" and "Next" buttons. They could serve either of two functions: 
- Move to previous/next item in a playlist 
- Move to previous/next chapter 

It's unlikely that a single player would need both, and meeting that need is offset by the need to avoid cluttering the interface. 

Instead, the previous/next buttons could be configurable via a **data-prevnext-unit** attribute, with supported values of "playlist" and "chapter". If this attribute is omitted, the buttons are omitted. 

The question then becomes: What should these buttons look like? Some specific questions:
- Buttons with text "Prev" and "Next" would work well in English, but how well do they localize to other languages? 
- Buttons with arrows might work better than text, but what would these look like? How would they be different than the current Rewind/Fast Forward buttons? 
- Do we still need rewind/fast forward buttons?  Or does the accessible seekbar provide sufficient means for scrubbing forward and backward? If these buttons are unnecessary, perhaps we could repurpose them as Previous/Next buttons. 

In any case, a title attribute (and an aria-label) could help to clarify the function of these buttons. 

Thoughts?

**Comment by Egyrs** [comment]
About localization: In Spanish they should be: "Prev" -> "Anterior" & "Next" -> "Siguiente".
The "prev" and "next" buttons can and must be different from the rewind/fast buttons:
[https://www.iconfinder.com/icons/22935/stepforwardnormal_icon#size=128](url) for next
[http://www.free-icons-download.net/images/previous-button-13045.png](url) for previous
Note that they have a bar before the arrow.
I think that is not a good idea to repurpose the buttons existing now, because theire meaning is different.
I'm not sure the seekbar can replace entirely the buttons fast forward and rewind. It should be done with test users.

**Comment by kensgists** [comment]
I'm wondering if a slight re-think of the organization of top buttons would help clarify functionality and clean things up a bit, especially if adding further buttons is a possibility. Plyr (http://plyr.io) keeps the skip forward/back buttons, yet is less cluttered than the current Able Player button arrangement. Plyr replaces the Stop with a "Restart" button, putting it before the rewind. So the grouping is Restart, Rewind x seconds, Play (Pause), and Forward x seconds. This seems very logical and it clears up the issue with having the seek/progress bar bounded by rewind and forward buttons.

I do think the time skip buttons should be kept. It is pretty difficult to use the seek bar with speech recognition software. The skips are much easier.

As for skipping by chapters or through tracks in a playlist, I wouldn't add those to the toolbar at the base level. It just adds visual clutter. Instead, that functionality should be handled within the TOCs for each. It would be OK to put "skip to next track"/"skip to next chapter" buttons within those TOC popup groupings of controls, though it might just clutter them up. I guess what I'm getting at is that the track/chapter skips may just be visual/cognitive kruft. Are they needed at all? Is there a clear use case?

**Comment by terrill** [comment]
@kensgists I like having the Play button first since I think it's the most frequently used button, and most likely action a person wants to take when they arrive at a video. A lot of current players, including Plyr, YouTube, and several others, have a full-width seekbar that's separate from the button controls. What do you think of that approach? 

Regarding the use case for "Previous" and "Next" buttons, if a playlist or chapter menu is long and a keyboard user wants to jump from Item 1 to Item 10 and their focus is currently on the controller, it would be a lot easier to simply press "Next" 10 times, rather than navigate back to the list to select the desired item. Also, maybe they're not sure which item they want - they just want to sample a little bit of each then move on if it doesn't capture their interest.  

I like the icons you suggested @Egyrs - a single triangle positioned outside of a vertical line seems logical. It suggests moving beyond the start or end, depending on direction of the triangle. One other person suggested this same idea off-line. 

Also keep in mind that this would be entirely optional for the developer. These buttons would only be present if data-prevnext-unit is present (and is set to either "chapter" or "playlist")

**Comment by sniderl** [comment]
I would leave the time skip buttons as people do use them a lot. I would like to see the Prev and Back only appear if chapters/playlist is used-as sometimes these things aren't used (I don't use them myself, but I can see how they would be useful.

As the others have said, the placement is the key for sighted users-in terms of whether a simple icon makes sense in its location.

I hadn't seen Plyr, looks really clean, although it has fewer options and is less accessible for people with vision disabilities to the grey on white...

**Comment by terrill** [comment]
Hey! I just noticed that YouTube now has a "Next" button to jump to the next item in a playlst. It appears on the far left, adjacent to the Play button. The icon they're using is like the one you proposed @Egyrs 
![youtube](https://cloud.githubusercontent.com/assets/912874/15031261/3968bd4a-120c-11e6-8c2b-6015830ecc69.png)

**Comment by wnereiz** [comment]
I don't think Rewind/Fast Forward button are useful any more. I agree to replace them to Prev/Next button. This is more reasonable. And I think it's not difficult.
The button style could refer to youtube, but maybe we can make a little different - like >>|

**Comment by yknivag** [comment]
Is there any progress on providing next/previous buttons?

Able player is one of the most accessible options I've managed to find, but no next/previous support means I can't use it.  This is a shame as the site I'm developing is exclusively aimed at visually impaired persons and so must be screen-reader compatible and there aren't man options...

**Comment by terrill** [comment]
@yknivag this is still tagged as "question", not yet on our todo list. My current thinking, based in part on evidence gathered through user testing, is: 

- We need to keep the rewind/fast forward buttons. The seek bar alone is not sufficient for all users.
- Chapters are not yet widely used. Therefore, there's not a high demand for buttons to easily navigate to the previous or next chapter. For media that does have chapters, the chapter menu provides this functionality. 
- Therefore, Previous and Next buttons could be added solely to navigate between items in a playlist. 
- The iconography for Prev/Next is fairly well established thanks to other players that support this: 
|< for Previous 
>| for Next

The current barrier to implementation though is that we haven't yet determined a standard format for playlists. So, adding these buttons is dependent on resolving the "Improve Playlist Support" issue: #157, which is a high priority.   

Thanks for letting us know how important this issue is to you. That will help us to prioritize.

**Comment by yknivag** [comment]
Hi @terrill many thanks for getting back so quickly!

I agree wholeheartedly with your comments, they correlate directly to my extensive searching elsewhere.  I hope you can resolve the connected issue and then this one.  If there is anything I can do to help then please let me know.

**Comment by terrill** [comment]
FYI - Now that playlists are more fully supported (as of the [version 4.0 release](https://github.com/ableplayer/ableplayer/releases/tag/v4.0)), I plan to get back to this issue soon, motivated in part by a nudge from @lardconcepts.

**Comment by yknivag** [comment]
@terrill Thank you for your continued dedication to this project and for not forgetting this issue.

I look forward to seeing how you solve it.

**Comment by lardconcepts** [comment]
+1 in agreement with @yknivag - great project.

Regarding solving the playlist problem,

I also tried setting data-seek-interval="10000" which kind of works, but with a long track and slow internet, everything hangs while the media loads so it can seek to the end, before it plays the next track. Which is surprising as the server accepts byte-ranges.

I did find a hack that would use jQuery to "click" the next item in the playlist after the active item, but of course, that brought focus away from where the user thought they currently were.
I know I should be able to "talk to" AblePlayer via $(AblePlayer)[0]
and $(".able-playlist"), but I failed.  My jQuery is not that strong!

Terril kindly said he'd look into around August, however, I need to hack SOMETHING together in days, to demo it as a possible solution for blind people. But without track list advance, it's a non-starter.

I'm happy to hack away at it myself, and submit a PR if I get anywhere, but for a short term "quick fix", if anyone can point me to a jQuery idea, I am happy to just implement a button next to the player for now.
Doesn't have to be perfect, anything is better than nothing for now!


Thanks!

**Comment by lardconcepts** [comment]
Right, I was getting desperate to launch, so I sat up one night, learnt enough jQuery to hack this together! It's good enough for now, and bring the focus back to where it was (which was what was stopping me before). For anyone who needs track controls (which I guess is everyone!) give this a go.

In the head, after jQuery is loaded:
```
  <script>
            $(function () {
                $('#ontrack').click(function () {
                    $('.able-current').next().click();
                   $('#ontrack').focus();
                });
                $('#backtrack').click(function () {
                    $('.able-current').prev().click();
                     $('#backtrack').focus();
                });
            });
        </script>
```

Just above the player:

```
<button id="ontrack" class="pure-button pure-button-active">Next Track</button>
<button id="backtrack" class="pure-button pure-button-active">Previous Track</button>
```

**Comment by terrill** [comment]
Previous and Next buttons are now added to the controller if a playlist is present. 
Implemented in db8821eed54ddb93a8b8c1e8c9f332c33d9c26d4

Thanks everyone for your patience!
